### ISSUE TITLE
ssh tunnling and How to use it.

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#62]
![How ssh tunnel and reverse portal forwarding working](https://iximiuz.com/ssh-tunnels/ssh-tunnels.png)

## Local Port Forwarding: 
```
ssh -L [from-start]<port>:[to-end]<port> user@sshd_server
```

Suppose you are try to reach a computer at forward-end. But on you local machine, you can only access forward-start, then you need to forward the port from start-port to the end-point (which is out of your reach without sshd)

![Port forwarding](https://iximiuz.com/ssh-tunnels/local-port-forwarding-bastion-2000-opt.png)

## Remote(Reverse) Port Forwarding: 
Here is what remote port forwarding can be used for:

- Exposing a dev service from your laptop to the public Internet for a demo.

When you want to momentarily expose a local service to the outside world,  you need a public-facing ingress gateway server. Any public-facing server with an SSH daemon (sshd) on it can be used as such a gateway:

```
ssh -R [from-start]<port>:[to-end]<port> user@sshd_server

ssh -R [remote_addr]:<remote_port>:[local_addr]:<local_port> [user@]gateway_addr

```
![Port forwarding](https://iximiuz.com/ssh-tunnels/local-port-forwarding-bastion-2000-opt.png)

By default, the above SSH tunnel will allow using only the gateway's localhost as the remote address.
In other words, your local port will become accessible only from the inside of the gateway server itself, and highly likely it's not something you actually need. For instance, I typically want to use the gateway's public address as the remote address to expose my local services to the public Internet. For that, the SSH server needs to be configured with the [GatewayPorts yes](https://linux.die.net/man/5/sshd_config#GatewayPorts) setting.

Pro Tip: Use `ssh -f -N -R` to run the port-forwarding session in the background.

# References
https://iximiuz.com/en/posts/ssh-tunnels/
